src/main/java/com/iota/iri/Snapshot.java

```diff
@@ -2,175 +2,181 @@
 
 import com.iota.iri.controllers.TransactionViewModel;
 import com.iota.iri.model.Hash;
-import org.apache.commons.io.IOUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.*;
+import java.util.Collections;
 import java.util.HashMap;
-import java.util.Iterator;
 import java.util.Map;
+import java.util.concurrent.Callable;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
-import java.util.stream.Collectors;
+import java.util.function.Supplier;
 
 
-public class Snapshot {
+public final class Snapshot {
+
     private static final Logger log = LoggerFactory.getLogger(Snapshot.class);
+
     public static String SNAPSHOT_PUBKEY = "TTXJUGKTNPOOEXSTQVVACENJOQUROXYKDRCVK9LHUXILCLABLGJTIPNF9REWHOIMEUKWQLUOKD9CZUYAC";
     public static int SNAPSHOT_PUBKEY_DEPTH = 6;
-    public static int SNAPSHOT_INDEX = 2;
+    private static int SNAPSHOT_INDEX = 2;
     public static int SPENT_ADDRESSES_INDEX = 3;
-    private static Snapshot initialSnapshot;
-
-
-    public final ReadWriteLock rwlock = new ReentrantReadWriteLock();
-
-
-    public static Snapshot init(String snapshotPath, String snapshotSigPath, boolean testnet) {
-        //This is not thread-safe (and it is ok)
-        if (initialSnapshot == null) {
-            if (!testnet && !SignedFiles.isFileSignatureValid(snapshotPath, snapshotSigPath, SNAPSHOT_PUBKEY,
-                    SNAPSHOT_PUBKEY_DEPTH, SNAPSHOT_INDEX)) {
-                throw new RuntimeException("Snapshot signature failed.");
-            }
-            Map<Hash, Long> initialState = initInitialState(snapshotPath);
-            initialSnapshot = new Snapshot(initialState, 0);
-            checkStateHasCorrectSupply(initialState);
-            checkInitialSnapshotIsConsistent(initialState);
 
-        }
-        return initialSnapshot;
-    }
+    private final ReadWriteLock rwlock = new ReentrantReadWriteLock();
+    private final boolean immutable;
+    private final Map<Hash, Long> state;
+    private int index;
 
```

Comment:
Most of the changes to Snapshot ensure that (1) state objects are never exposed (copied if needed elsewhere) and (2) that locking is internalized with access provided via lambda accepting methods.
Also, immutable flag is used to make Snapshot from static `init` method immutable.